pkg/yttlibrary/overlay/array.go

```diff
@@ -4,13 +4,18 @@
 package overlay
 
 import (
+	"github.com/k14s/ytt/pkg/template"
 	"github.com/k14s/ytt/pkg/yamlmeta"
 )
 
 func (o Op) mergeArrayItem(
 	leftArray *yamlmeta.Array, newItem *yamlmeta.ArrayItem,
 	parentMatchChildDefaults MatchChildDefaultsAnnotation) error {
 
+	if !template.NewAnnotations(newItem).Has(AnnotationMatch) {
```

Comment:
what do you think encapsulating interpretation of overlay/match for array items within NewArrayItemMatchAnnotation class. it seems that it already does a similar check here: https://github.com/vmware-tanzu/carvel-ytt/blob/767de6eada88abc6720dcb2422e6b2f6586f9bc2/pkg/yttlibrary/overlay/array_item_match_annotation.go#L36-L39

Reply:
Great suggestion. I changed `NewArrayItemMatchAnnotation` to return a typed error if a `overlay/match` is not present, and then have `array.go` check the error type. Let me know if that doesn't capture the encapsulation you were thinking of.

Reply:
i was hinting at to go even further. currently you have a special conditional that short circuits and does the append if overlay/match isnt defined. NewArrayItemMatchAnnotation encapsulates all of the thinking behind how to interpret overlay/match and currently errors if it's not provided. instead of catching that error, we can expand NewArrayItemMatchAnnotation to note that overlay/match was not provided and hence return 0 matches from MatchNodes function [0]. this would keep all the matching logic within that class and would rely on existing check for 0 matched nodes to do appending [1]

[0] https://github.com/vmware-tanzu/carvel-ytt/pull/301/files#diff-880c2538f7db2e19b5ac04ffbc8f4a059e07cd8a6a1444bd0228699d6cdc83ccR87

[1] https://github.com/vmware-tanzu/carvel-ytt/pull/301/files#diff-1ec3c53bb1375f91da8a493c974f9f8f64dd1a3c1a3cc19059ec04bb395ca247R35-R37

Reply:
Steven and I are looking into this and we found that if you misspell `#@overlay/match` (making `if !anns.Has(AnnotationMatch)` true) the array item will be default. However we feel like this should be an error case, rather than a default append case.

Reply:
^ this is probably out of scope for this issue. typo-ing annotations is a possibility for all annotations so solution needs to be holistic. related issue: https://github.com/vmware-tanzu/carvel-ytt/issues/114